### PR TITLE
Add a new structure for rheology models

### PR DESCRIPTION
**Pull request by gassmoeller**
This is a first stab at the new structure for outsourcing and deduplicating viscosity computations between material models. I am grateful for comments.

### CODE REVIEW

File: include/aspect/material_model/rheology/constant.h [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+#include <aspect/global.h>
+
+namespace aspect
+{
+  namespace MaterialModel
+  {
+    using namespace dealii;
+
+    namespace Rheology
```
Comment (MFraters) [inline include/aspect/material_model/rheology/constant.h]: RheologyModel?

**Comment by naliboff** [comment]
@gassmoeller - Thank for you taking the first crack at this! 

A general comment about the naming schemes for different properties. Here, you used `Constant` to represent a constant viscosity. 

However, there are potential use cases (ex: viscoelasticity) where the value being returned is a force or stress. Another example is grain-size - would there be a separate function to compute grain size or should this be done in the material model. Yet another use case is strain-softening, but perhaps this should go in the `MaterialUtilities`.

If the classes under `Rheology` are really only designed to return viscosities, should the namespace be changed to `Viscosity`? For me, "rheology" implies anything related to the constitutive behavior (stress, strain, viscosity, etc).

File: include/aspect/material_model/rheology/constant.h (hunk 1 above)
Comment (gassmoeller) [reply include/aspect/material_model/rheology/constant.h]: Personally I want to get rid of as many `model` as possible (we renamed all the boundary conditions and initial conditions for 2.0). I can live with `MaterialModel`, but I do not want to introduce more of them :smile:.

**Comment by gassmoeller** [comment]
I think I addressed all comments, and also mentioned in the changelog that the deduplication changed a default value.

**Comment by gassmoeller** [comment]
Yes my idea was to keep the function as simple as possible and not worry about an interface.

File: include/aspect/material_model/rheology/constant_viscosity.h [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+          static
+          void
+          declare_parameters (ParameterHandler &prm);
+
+          /**
+           * Read the parameters from the parameter file.
+           */
+          virtual
+          void
+          parse_parameters (ParameterHandler &prm);
```
Comment (bangerth) [inline include/aspect/material_model/rheology/constant_viscosity.h]: If you make the function `virtual`, then that is only useful if you want to use the class in a class hierarchy. But then you also want a virtual destructor.

In the end, I think you probably don't want that, and could just remove the `virtual`.

File: include/aspect/material_model/rheology/constant_viscosity.h [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+          /**
+           * Compute the viscosity, in this case just a constant.
+           */
+          double
+          compute_viscosity () const;
+
+        private:
+
+          double eta;
```
Comment (bangerth) [inline include/aspect/material_model/rheology/constant_viscosity.h]: Maybe a one-line comment? For example how this variable is initialized?

Comment (bangerth) [reply include/aspect/material_model/rheology/constant_viscosity.h]: The class could use a constructor that sets the variable to an invalid value.

File: source/material_model/rheology/constant_viscosity.cc [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+      {
+        return eta;
+      }
+
+
+
+      void
+      ConstantViscosity::declare_parameters (ParameterHandler &prm)
+      {
+        prm.declare_entry ("Viscosity", "1e21",
```
Comment (bangerth) [inline source/material_model/rheology/constant_viscosity.cc]: What was the default value before and why can't we use it?

File: include/aspect/material_model/rheology/constant_viscosity.h [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    {
+      class ConstantViscosity
+      {
+        public:
+          /**
+           * Declare the parameters this function takes through input files.
+           */
+          static
+          void
+          declare_parameters (ParameterHandler &prm);
```
Comment (bangerth) [inline include/aspect/material_model/rheology/constant_viscosity.h]: You could avoid the problem with the change of default if you let this function take an argument for the constant viscosity. This way, each model that uses this rheology could just use its own preferred value.

**Comment by gassmoeller** [comment]
I addressed the comments and added a optional argument that allows to choose the default viscosity. Ready from my side.

**Pull request merged**